README.md
# Very short description of the package

[![Latest Version on Packagist](https://img.shields.io/packagist/v/savannabits/advantasms.svg?style=flat-square)](https://packagist.org/packages/savannabits/advantasms)
[![Build Status](https://img.shields.io/travis/savannabits/advantasms/master.svg?style=flat-square)](https://travis-ci.org/savannabits/advantasms)
[![Quality Score](https://img.shields.io/scrutinizer/g/savannabits/advantasms.svg?style=flat-square)](https://scrutinizer-ci.com/g/savannabits/advantasms)
[![Total Downloads](https://img.shields.io/packagist/dt/savannabits/advantasms.svg?style=flat-square)](https://packagist.org/packages/savannabits/advantasms)

The PHP SDK for the [AdvantaSMS BulkSMS API](https://advantasms.com/bulksms-api)
## Requirements
- PHP ^7.1
- PHP json extension
- PHP Curl extension

## Installation

You can install the package via composer:

```bash
composer require savannabits/advantasms
```

## Usage
### Credentials
Ensure you have thhe following required credentials from AdvantaSMS:
 - apiKey
 - Partner ID (e.g 2030)
 - Shortcode (e.g SAVBITS)
 
### Sending SMS
```php
use \Savannabits\Advantasms\Advantasms;
$apiKey = "";
$partnerId = "";
$shortcode = "";
$mobile = "254xxxxxxxxx";
//instantiate
$sms = new Advantasms($apiKey,$partnerId,$shortcode);

//Send and receive response
$response = $sms->to($mobile)->message("Your message right here...")->send();

//Schedule sms to be sent at a specific time
$time = "2020-10-01 18:00"; // Y-m-d H:i
$response = $sms->to($mobile)->message("Your message right here")->schedule($time);
```

If you don't like instantiating class into variables, you can use the init static method instead:
```php
use \Savannabits\Advantasms\Advantasms;
$apiKey = "";
$partnerId = "";
$shortcode = "";
$mobile = "254xxxxxxxxx";

//Send and receive response
$response = Advantasms::init($apiKey,$partnerId,$shortcode)->to($mobile)->message("Your message right here...")->send();

//Schedule sms to be sent at a specific time
$time = "2020-10-01 18:00"; // Y-m-d H:i
$response = Advantasms::init($apiKey,$partnerId,$shortcode)->to($mobile)->message("Your message right here...")->schedule($time);
```

### Sample responses

#### Error
```json
{
  "success":false,
  "message":"Low credit units to send message, Current balance 0.00, Required 1",
  "payload":[
        {
          "response-code":1004,
          "response-description":"Low credit units to send message, Current balance 0.00, Required 1",
          "mobile":"0708467001"
        }
     ],
  "code":1004
}
```
#### Success
```json
{
  "success":true,
  "message":"Success",
  "payload":[
        {
            "respose-code": 200,
            "response-description": "Success",
            "mobile": 254712345678,
            "messageid": 8290842,
            "networkid": "1"
        }
    ],
  "code":200
}
```


### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email maosa.sam@gmail.com instead of using the issue tracker.

## Credits

- [Sam Maosa](https://github.com/savannabits)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `advantasms` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/ExampleTest.php
<?php

namespace Savannabits\Advantasms\Tests;

use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


src/Advantasms.php
<?php

namespace Savannabits\Advantasms;

class Advantasms
{
    private $apikey, $partnerId, $shortcode;
    private $message, $to;
    private $baseUrl = "https://quicksms.advantasms.com";
    private $sendsms = "/api/services/sendsms/";
    /**
     * Advantasms constructor.
     * @param string $apiKey |The advanta sms API Key. See documentation for more details
     * @param string $partnerId | The Partner ID. See advantaSMS documentation for more details
     * @param string $shortCode | The Shortcode of used to send sms. See documentation for more details
     * @param string|null $domain | The base domain in case it is different from https://quicksms.advantasms.com
     * @return Advantasms
     */
    public function __construct($apiKey, $partnerId, $shortCode, $baseUrl="https://quicksms.advantasms.com")
    {
        $this->apikey = $apiKey;
        $this->partnerId = $partnerId;
        $this->shortcode = $shortCode;
        $this->baseUrl  = $baseUrl;
        return $this;
    }

    /**
     * Instantiate the Advantasms class.
     * @param string $apiKey |The advanta sms API Key. See documentation for more details
     * @param string $partnerId | The Partner ID. See advantaSMS documentation for more details
     * @param string $shortCode | The Shortcode of used to send sms. See documentation for more details
     * @param string|null $domain | The base domain in case it is different from quicksms.advantasms.com
     * @return Advantasms
     */
    public static function init($apiKey,$partnerId, $shortCode, $baseUrl="https://quicksms.advantasms.com") {
        $instance = new self($apiKey,$partnerId,$shortCode,$baseUrl);
        return $instance;
    }

    /**
     * @param $mobileNumber
     * @return Advantasms
     */
    public function to($mobileNumber) {
        $this->to = $mobileNumber;
        return $this;
    }

    /**
     * @param string $message
     * @return Advantasms
     */
    public function message(string $message="") {
        $this->message = $message;
        return $this;
    }

    /**
     * Execute sms sending action
     * @return array|mixed
     *
     * 200;Successful Request Call
     * 1001;Invalid sender id
     * 1002;Network not allowed
     * 1003;Invalid mobile number
     * 1004;Low bulk credits
     * 1005;Failed. System error
     * 1006;Invalid credentials
     * 1007;Failed. System error
     * 1008;No Delivery Report
     * 1009;unsupported data type
     * 1010;unsupported request type
     * 4090;Internal Error. Try again after 5 minutes
     * 4091;No Partner ID is Set
     * 4092;No API KEY Provided
     * 4093;Details Not Found
     * */
    public function send() {
        $data = [
            "apikey"=>$this->apikey,
            "partnerID"=>trim($this->partnerId),
            "message"=>trim($this->message),
            "shortcode"=>$this->shortcode,
            "mobile"=>trim($this->to),
            'pass_type' => 'plain',
        ];
        $response = $this->curlPost($this->sendsms,$data);
        $return = [
            "success" => false,
            "message" => "",
            "payload" => []
        ];
        if (!$response) {
            $return["success"] = false;
            $return["message"] = "No response from the server.";
            return $return;
        } else {
            if (isset($response['responses'])) {
                $first = $response["responses"][0];
                $return["success"] = $first["response-code"] ===200;
                $return["code"] = $first["response-code"];
                $return["message"] = $first["response-description"];
                $return["payload"] = $response["responses"];
                return $return;
            }
            if (isset($response["response-code"])) {
                $first = $response;
                $return["success"] = $first["response-code"] ===200;
                $return["code"] = $first["response-code"];
                $return["message"] = $first["response-description"];
                $return["payload"] = $response;
                return $return;
            }
            //Temporal fix for the mis-spelled api response code to 'respose-code'
            if (isset($response["respose-code"])) {
                $first = $response;
                $return["success"] = $first["respose-code"] ===200;
                $return["code"] = $first["respose-code"];
                $return["message"] = $first["response-description"];
                $return["payload"] = $response;
                return $return;
            }

            $return["success"] = false;
            $return["message"] = "Unknown Error";
            $return["payload"] = $response;
            return $return;
        }
    }

    /**
     * Schedule sms sending action
     * @param string $time | Time to send in Y-m-d H:i format
     * @return array|mixed
     *
     * 200;Successful Request Call
     * 1001;Invalid sender id
     * 1002;Network not allowed
     * 1003;Invalid mobile number
     * 1004;Low bulk credits
     * 1005;Failed. System error
     * 1006;Invalid credentials
     * 1007;Failed. System error
     * 1008;No Delivery Report
     * 1009;unsupported data type
     * 1010;unsupported request type
     * 4090;Internal Error. Try again after 5 minutes
     * 4091;No Partner ID is Set
     * 4092;No API KEY Provided
     * 4093;Details Not Found
     * */
    public function schedule($time) {
        $data = [
            "apikey"=>$this->apikey,
            "partnerID"=>trim($this->partnerId),
            "message"=>trim($this->message),
            "shortcode"=>$this->shortcode,
            "mobile"=>trim($this->to),
            "timeToSend" => trim($time),
            'pass_type' => 'plain',
        ];
        $response = $this->curlPost($this->sendsms,$data);
        $return = [
            "success" => false,
            "message" => "",
            "payload" => []
        ];
        if (!$response) {
            $return["success"] = false;
            $return["message"] = "No response from the server.";
            return $return;
        } else {
            if (isset($response['responses'])) {
                $first = $response["responses"][0];
                $return["success"] = $first["response-code"] ===200;
                $return["code"] = $first["response-code"];
                $return["message"] = $first["response-description"];
                $return["payload"] = $response["responses"];
                return $return;
            }
            if (isset($response["response-code"])) {
                $first = $response;
                $return["success"] = $first["response-code"] ===200;
                $return["code"] = $first["response-code"];
                $return["message"] = $first["response-description"];
                $return["payload"] = $response;
                return $return;
            }
            //Temporal fix for the mis-spelled api response code to 'respose-code'
            if (isset($response["respose-code"])) {
                $first = $response;
                $return["success"] = $first["respose-code"] ===200;
                $return["code"] = $first["respose-code"];
                $return["message"] = $first["response-description"];
                $return["payload"] = $response;
                return $return;
            }

            $return["success"] = false;
            $return["message"] = "Unknown Error";
            $return["payload"] = $response;
            return $return;
        }
    }
    /**
     * @param string $endpoint
     * @param array $data
     * @param array $headers
     * @return array|mixed
     */
    private function curlPost(string $endpoint, array $data, array $headers=[]) {
        $url = $this->baseUrl.$endpoint;
        $curl = curl_init();
        curl_setopt($curl, CURLOPT_URL, $url);
        curl_setopt($curl, CURLOPT_HTTPHEADER, array_merge(array('Content-Type:application/json'),$headers));

        $data_string = json_encode($data);
        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($curl, CURLOPT_POST, true);
        curl_setopt($curl, CURLOPT_POSTFIELDS, $data_string);
        curl_setopt($curl, CURLOPT_HEADER, false);
        $curl_response = curl_exec($curl);
        return json_decode($curl_response,true);
    }

}
